app/Dockerfile

```diff
@@ -2,29 +2,57 @@
 # https://hub.docker.com/_/python
 
 # The build stage that will be used to deploy to the various environments
-# needs to be called `release` in order to integrate with the repo's
-# top-level Makefile
-FROM python:3-slim AS release
-
-# Install poetry, the package manager.
-# https://python-poetry.org
-RUN pip install poetry
+# This is what the other stages will call as their base
+FROM python:3.11-slim-bullseye AS build
 
+# Keep container packages up-to-date.
 RUN apt-get update \
     && apt-get install --no-install-recommends --yes \
     build-essential \
     libpq-dev \
     postgresql
 
+# Install poetry, the package manager - https://python-poetry.org
+# We use --no-cache-dir here because the pip caches normally, and if 
+# we don't cache, it will reduce the overall image size
+# The version is pinned so that there is a known good configuration 
+# that we have better control over vs a range that can be used
+RUN pip install --no-cache-dir poetry==1.3.1
+
+# setup user stuff
+ARG RUN_UID
+ARG RUN_USER
+
+# The following logic creates the RUN_USER home directory and the directory where
+# we will be storing the application in the image. This runs when the user is not root
+RUN : "${RUN_USER:?RUN_USER and RUN_UID need to be set and non-empty.}" && \
+    [ "${RUN_USER}" = "root" ] || \
+    (useradd -mU --home "/home/${RUN_USER}" --uid ${RUN_UID} "${RUN_USER}" \
+    && mkdir /app \
+    && chown -R ${RUN_UID} "/home/${RUN_USER}" /app)
+
+USER ${RUN_USER}
+
+#-------------------------------------------------------------------------------
+# Development build environment
+#-------------------------------------------------------------------------------
+FROM build as dev
+
+# Runtime arguement for what user will the bulk of this stage run as
+ARG RUN_USER
+
+# In between ARG RUN_USER and USER ${RUN_USER}, the user is still root
+# If there is anything that needs to be ran as root, this is the spot
+
+# Change the user from root to ${RUN_USER}
+USER ${RUN_USER}
+
 # Set the application working directory.
-WORKDIR /srv
+WORKDIR /app
 
-COPY pyproject.toml poetry.lock ./
+COPY . /app
```

Comment:
There could be a slight performance gain since it would be cached, but we wouldn't be updating the `pyproject` or `poetry` files super often, since we should be using `dependabot` to update these files. I can split it back up, but IDK if it will save much time overall